keps/sig-network/3705-cloud-node-ips/README.md

```diff
@@ -0,0 +1,687 @@
+IP request can't be fulfilled, log an error, and leave the node in the
+tainted state.
+
+It makes sense to have the same behavior if the user passes a "new"
+(eg, dual-stack) `--node-ip` value to kubelet, but the cloud provider
+does not recognize the new syntax and thus can't fulfil the request.
+Conveniently, we can do this just by passing the dual-stack
+`--node-ip` value in the existing annotation; the old cloud provider
+will try to parse it as a single IP address, fail, log an error, and
+leave the node in the tainted state, which is exactly what we wanted
+it to do if it can't interpret the `--node-ip` value correctly.
+
+Therefore, we do not need a new annotation for the new `--node-ip`
+values; we can continue to use the existing annotation, assuming
+existing cloud providers will treat unrecognized values as errors.
+
+```
+<<[UNRESOLVED annotation-name ]>>
+
+The annotation name is `alpha.kubernetes.io/provided-node-ip`
+but it hasn't been "alpha" for a long time. Should we rename it? In
+that case, we probably need to keep supporting both versions for a
+while.
+
+<<[/UNRESOLVED]>>
+```
+
+Kubelet will preserve the existing behavior of _not_ passing
+"`0.0.0.0`" or "`::`" to the cloud provider, for backward
+compatibility, but it _will_ pass "`IPv4`" and "`IPv6`" if they are
+passed as the `--node-ip`.
+
+### Changes to cloud providers
+
+Assuming that all cloud providers use the `"k8s.io/cloud-provider"`
+code to handle the node IP annotation and node address management, no
+cloud-provider-specific changes should be needed; we should be able to
+make the needed changes in the `cloud-provider` module, and then the
+individual providers just need to revendor to the new version.
+
+### Example of `--node-ip` possibilities
+
+Assume a node where the cloud has assigned the IPs `1.2.3.4`,
+`5.6.7.8`, `abcd::1234` and `abcd::5678`, in that order of preference.
+
+("SS" = "Single-Stack", "DS" = "Dual-Stack")
+
+| `--node-ip` value    | New? | Annotation             | Resulting node addresses |
+|----------------------|------|------------------------|--------------------------|
+| (none)               | no   | (unset)                | `["1.2.3.4", "5.6.7.8", "abcd::1234", "abcd::5678"]` (DS IPv4-primary) |
+| `0.0.0.0`            | no   | (unset)                | `["1.2.3.4", "5.6.7.8", "abcd::1234", "abcd::5678"]` (DS IPv4-primary) |
+| `::`                 | no   | (unset)                | `["1.2.3.4", "5.6.7.8", "abcd::1234", "abcd::5678"]` (DS IPv4-primary *) |
+| `1.2.3.4`            | no   | `"1.2.3.4"`            | `["1.2.3.4"]` (SS IPv4) |
+| `9.10.11.12`         | no   | `"9.10.11.12"`         | (error, because the requested IP is not available) |
+| `abcd::5678`         | no   | `"abcd::5678"`         | `["abcd::5678"]` (SS IPv6) |
+| `1.2.3.4,abcd::1234` | yes* | `"1.2.3.4,abcd::1234"` | `["1.2.3.4", "abcd::1234"]` (DS IPv4-primary) |
+| `IPv4`               | yes  | `"IPv4"`               | `["1.2.3.4"]` (SS IPv4) |
```

Comment:
Why would this be a single IP and not all IPv4 ?

Same for all subsequent

Reply:
I thought "single IP" was more consistent with the case of specifying a single IP (where doing so eliminates all the other IPs of that family). As I defined it, `IPv4` meant "pick an IPv4 address", not "pick _every_ IPv4 address"...

Reply:
I'm not confident in my opinion either way on this. How can we get some feedback on this point?